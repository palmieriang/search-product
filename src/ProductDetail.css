.detail-container {
  width: 30rem;
  margin-left: 32px;
  padding: 0 16px;
  background: #ffffff;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.2);
}

.row {
  display: flex;
  margin: 16px 0;
}

.row-title {
  width: 8rem;
  flex-shrink: 0;
  font-weight: bold;
}

.product-image {
  max-width: 100%;
}
