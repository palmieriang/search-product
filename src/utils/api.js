export const fetchSuggestions = (searchTerm) => {
  return fetch(
    "https://asoscarewebinterviewstub.azurewebsites.net/api/search",
    {
      method: "POST",
      body: JSON.stringify({
        q: searchTerm,
      }),
    }
  ).then((res) => res.json());
};

export const fetchProductDetail = (id) => {
  return fetch(
    "https://asoscarewebinterviewstub.azurewebsites.net/api/product",
    {
      method: "POST",
      body: JSON.stringify({
        id,
      }),
    }
  ).then((res) => res.json());
};
