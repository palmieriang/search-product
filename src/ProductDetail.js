import React, { useState, useEffect } from "react";

import { fetchProductDetail } from "./utils/api";

import "./ProductDetail.css";

const ProductDetail = ({ productId }) => {
  const [productInfo, setProductInfo] = useState(null);

  useEffect(() => {
    if (!productId) return;

    fetchProductDetail(productId).then((productInfo) =>
      setProductInfo(productInfo)
    );
  }, [productId]);

  const renderProductInfo = () => {
    return (
      <div className="detail-container">
        <div className="row">
          <img src={productInfo.productImage} className="product-image" />
        </div>
        <div className="row">
          <div className="row-title">Name:</div>
          <div className="row-body">{productInfo.name}</div>
        </div>

        <div className="row">
          <div className="row-title">Product Code:</div>
          <div className="row-body">{productInfo.productCode}</div>
        </div>

        <div className="row">
          <div className="row-title">Price:</div>
          <div className="row-body">{productInfo.price}</div>
        </div>
      </div>
    );
  };

  return productInfo && renderProductInfo();
}

export default ProductDetail;
