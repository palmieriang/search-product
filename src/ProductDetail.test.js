import React from "react";
import { shallow } from "enzyme";
import {render, screen, waitFor} from "@testing-library/react";

import ProductDetail from "./ProductDetail";
import { fetchProductDetail } from "./utils/api";

jest.mock("./utils/api");

const productId = '1790102';
const productInfo = {
  name: "shirt in navy",
  productCode: "1774327",
  productImage: "https://asos.com/image.json",
  price: "Â£60.00"
};

describe("ProductDetail", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<ProductDetail />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should show product id, name and price", async () => {
    fetchProductDetail.mockResolvedValueOnce(productInfo);

    render(<ProductDetail productId={productId} />);

    await waitFor(() => expect(screen.getByText(productInfo.name)).toBeInTheDocument());
    expect(screen.getByText(productInfo.productCode)).toBeInTheDocument();
    expect(screen.getByText(productInfo.price)).toBeInTheDocument();
    expect(screen.getByAltText(productInfo.name)).toHaveAttribute('src', productInfo.productImage);
  });
});
