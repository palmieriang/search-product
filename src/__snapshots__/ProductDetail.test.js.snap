// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail Renders correctly 1`] = `""`;

exports[`ProductDetail renders correctly 1`] = `""`;
